src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs

```diff
@@ -0,0 +1,175 @@
+                    return new KafkaHealthCheck(options);
+                },
+                failureStatus: default,
+                tags: default));
+        }
+    }
+
+    private static ConsumerConnectionFactory<TKey, TValue> CreateConsumerConnectionFactory<TKey, TValue>(IServiceProvider serviceProvider, Action<ConsumerBuilder<TKey, TValue>>? configureBuilder, KafkaConsumerSettings settings)
+        => new(CreateConsumerBuilder(serviceProvider, configureBuilder, settings), settings.Config!);
+    
+    private static ConsumerBuilder<TKey, TValue> CreateConsumerBuilder<TKey, TValue>(IServiceProvider serviceProvider, Action<ConsumerBuilder<TKey, TValue>>? configureBuilder, KafkaConsumerSettings settings)
+    {
+        settings.Validate();
+
+        ConsumerBuilder<TKey, TValue> builder = new(settings.Config);
+        ILogger logger = serviceProvider.GetRequiredService<ILoggerFactory>().CreateLogger(ConfluentKafkaCommon.LogCategoryName);
+        configureBuilder?.Invoke(builder);
+
+        try
+        {
+            void OnLog(IConsumer<TKey, TValue> _, LogMessage logMessage) =>
+                logger.Log((LogLevel)logMessage.LevelAs(LogLevelType.MicrosoftExtensionsLogging), 0, logMessage.Message, null, static (value, ex) => value);
+
+            builder.SetLogHandler(OnLog);
+        }
+        catch (InvalidOperationException)
+        {
+            logger.LogWarning("LogHandler is already set. Skipping... No logs will be written.");
+        }
+
+        if (settings.Metrics)
+        {
+            MetricsChannel channel = serviceProvider.GetRequiredService<MetricsChannel>();
+            void OnStatistics(IConsumer<TKey, TValue> _, string json)
+            {
+                if (string.IsNullOrEmpty(json))
+                {
+                    return;
+                }
+
+                // StatisticsHandler is called on the consume thread, so we need to offload the processing
+                // to avoid slowing the consumer down.
+                channel.Writer.TryWrite(json);
+            }
+
+            try
+            {
+                builder.SetStatisticsHandler(OnStatistics);
+            }
+            catch (InvalidOperationException)
+            {
+                logger.LogWarning("StatisticsHandler is already set. Skipping... No metrics will be exposed.");
+            }
+        }
+        return builder;
+    }
+
+    private static KafkaConsumerSettings BuildConsumerSettings(IHostApplicationBuilder builder, string configurationSectionName, Action<KafkaConsumerSettings>? configureSettings, string connectionName)
+    {
+        var configSection = builder.Configuration.GetSection(configurationSectionName);
+        KafkaConsumerSettings settings = new();
+        configSection.Bind(settings);
+
+        ConsumerConfig config = new();
+        configSection.GetSection(nameof(KafkaConsumerSettings.Config)).Bind(config);
+        settings.Config = config;
```

Comment:
Looking more at Kafka, there is a bit of a special pattern here in that the `ConsumerConfig` is separate from the `ConsumerBuilder`. We don't really have that pattern in the other components - typically there is just 1 object - ConfigurationOptions in Redis, IConnectionFactory in RabbitMQ.

Maybe what we have here is fine for now. It would be awkward to have 3 "configure" callbacks - configureSettings, configureConfig, configureBuilder.

Reply:
@eerhardt The reference to the underlying library configuration object in the Aspire's "Settings" object lets customer configure a dedicated serializer / deserialize per pair of generic parameters provided to the underlying confluent XXXBuilder. Using IOptions pattern here would defeat it.
I "reproduced" the pattern I observed in the RabbitMQ implementation.